src/lib.rs

```diff
@@ -1,307 +1,268 @@
-                msg::source(),
-                self.lottery_owner
-            );
+        ft_actor_id: Option<ActorId>,
+    ) -> GOCEvent {
+        self.assert_admin();
+
+        if self.started != 0 {
+            panic!("Current game round must be over");
         }
 
-        if self.lottery_is_on() {
-            self.players = BTreeMap::new();
+        if matches!(ft_actor_id, Some(ft_actor_id) if ft_actor_id.is_zero()) {
+            panic!("`ft_actor_id` mustn't be `ActorId::zero()`");
         }
 
-        self.lottery_started = true;
-        self.lottery_start_time = exec::block_timestamp();
-        self.lottery_duration = duration;
+        self.players.clear();
+
+        self.prize_fund = 0;
+        self.started = exec::block_timestamp();
+        self.ending = self.started + duration;
         self.participation_cost = participation_cost;
-        self.prize_fund = prize_fund;
-        self.token_address = token_address;
-        self.lottery_id = self.lottery_id.saturating_add(1);
-    }
+        self.ft_actor_id = ft_actor_id;
 
-    // checks that the player is already participating in the lottery
-    fn player_exists(&mut self) -> bool {
-        self.players
-            .values()
-            .any(|player| player.player_id == msg::source())
+        GOCEvent::Started {
+            ending: self.ending,
+            participation_cost,
+            ft_actor_id,
+        }
     }
 
-    /// Transfers `amount` tokens from `sender` account to `recipient` account.
-    /// Arguments:
-    /// * `from`: sender account
-    /// * `to`: recipient account
-    /// * `amount`: amount of tokens
-    async fn transfer_tokens(
-        token_address: &ActorId,
-        from: &ActorId,
-        to: &ActorId,
-        amount_tokens: u128,
-    ) {
-        msg::send_for_reply(
-            *token_address,
-            FTAction::Transfer {
-                from: *from,
-                to: *to,
-                amount: amount_tokens,
-            },
-            0,
-        )
-        .expect("Error in sending message")
-        .await
-        .expect("Error in transfer");
+    fn assert_admin(&self) {
+        if msg::source() != self.admin {
+            panic!("`msg::source()` must be the game administrator");
+        }
     }
 
-    /// Called by a player in order to participate in lottery
-    /// Requirements:
-    /// * Lottery must be on
-    /// * Contribution must be greater than zero
-    /// * The player cannot enter the lottery more than once
-    /// Arguments:
-    /// * `amount`: value or tokens to participate in the lottery
-    async fn enter(&mut self, amount: u128) {
-        if self.lottery_is_on() && !self.player_exists() && amount == self.participation_cost {
-            if let Some(ref token_address) = self.token_address {
-                Self::transfer_tokens(token_address, &msg::source(), &exec::program_id(), amount)
-                    .await;
+    async fn pick_winner(&mut self) -> GOCEvent {
```

Comment:
Perhaps here is a good example to show how delayed messages work
You can send the message `pick_winner` when starting a lottery

Reply:
But this is just an assumption because additional logic will be required to track messages to the fungible token contract